File: alaiy_os_core/api/connectors.py
import frappe


@frappe.whitelist()
def get_all_connectors():
    """Return all OS Connector Registry rows for the settings panel."""
    if not frappe.db.exists("DocType", "OS Connector Registry"):
        return []
    return frappe.get_all(
        "OS Connector Registry",
        fields=[
            "connector_id", "connector_name", "connector_app",
            "connector_type", "description", "icon", "icon_url",
            "settings_doctype", "test_method",
            "is_enabled", "connection_status", "last_tested_at",
        ],
        order_by="connector_name asc",
    )


@frappe.whitelist()
def get_connector_config(connector_id):
    """Return field metadata + current values for a connector's settings DocType."""
    registry = frappe.get_doc("OS Connector Registry", connector_id)
    settings_doctype = registry.settings_doctype

    if not settings_doctype:
        frappe.throw("No settings DocType configured for this connector.")
    if not frappe.db.exists("DocType", settings_doctype):
        frappe.throw(f"Settings DocType '{settings_doctype}' not found.")

    RENDERABLE = {"Data", "Password", "Int", "Link", "Select", "Check", "Text", "Small Text"}
    meta = frappe.get_meta(settings_doctype)
    fields = []
    for f in meta.fields:
        if f.fieldtype not in RENDERABLE:
            continue
        fields.append({
            "fieldname": f.fieldname,
            "label": f.label,
            "fieldtype": f.fieldtype,
            "options": f.options,
            "reqd": f.reqd,
            "description": f.description,
        })

    doc = frappe.get_single(settings_doctype)
    values = {}
    for f in fields:
        if f["fieldtype"] == "Password":
            raw = doc.get(f["fieldname"])
            values[f["fieldname"]] = {"_type": "password", "_set": bool(raw)}
        else:
            values[f["fieldname"]] = doc.get(f["fieldname"])

    return {"fields": fields, "values": values}


@frappe.whitelist()
def get_connector_password(connector_id, fieldname):
    """Return the decrypted value of a Password field for display."""
    registry = frappe.get_doc("OS Connector Registry", connector_id)
    meta = frappe.get_meta(registry.settings_doctype)
    field_meta = meta.get_field(fieldname)
    if not field_meta or field_meta.fieldtype != "Password":
        frappe.throw("Invalid field")
    doc = frappe.get_single(registry.settings_doctype)
    return doc.get_password(fieldname, raise_exception=False) or ""


@frappe.whitelist()
def save_and_test(connector_id, values):
    """Save connector settings then run the connector's test method."""
    import json

    if isinstance(values, str):
        values = json.loads(values)

    registry = frappe.get_doc("OS Connector Registry", connector_id)
    settings_doctype = registry.settings_doctype
    test_method = registry.test_method

    doc = frappe.get_single(settings_doctype)
    meta = frappe.get_meta(settings_doctype)

    for fieldname, value in values.items():
        field_meta = meta.get_field(fieldname)
        if not field_meta:
            continue
        if field_meta.fieldtype == "Password":
            if value and str(value).strip():
                doc.set(fieldname, value)
        else:
            doc.set(fieldname, value)

    doc.save(ignore_permissions=True)
    frappe.db.commit()

    try:
        test_fn = frappe.get_attr(test_method)
        result = test_fn()
    except Exception as e:
        result = {"success": False, "message": str(e)}

    registry.connection_status = "connected" if result.get("success") else "failed"
    registry.last_tested_at = frappe.utils.now_datetime()
    registry.save(ignore_permissions=True)
    frappe.db.commit()

    return result

File: interface/src/lib/frappe/connectors.ts
/**
 * The platform's connector-settings API, over `alaiy_os.api.connectors`.
 *
 * Those endpoints are generic over `OS Connector Registry`: they read the field
 * metadata and current values of whatever settings DocType a connector
 * registered, save them back, and run that connector's own `test_method`. So a
 * connector's settings screen needs no Python of its own, and the base needs no
 * knowledge of which connectors exist — the same registry-driven indirection
 * `alaiy_os/connectors.py` already uses to resolve sync methods.
 *
 * A connector's `interface/` imports this to build its own settings screen.
 */

export type ConnectorFieldType =
  | "Data"
  | "Password"
  | "Int"
  | "Float"
  | "Link"
  | "Select"
  | "Check"
  | "Text"
  | "Small Text"
  | "Section Break";

export interface ConnectorField {
  fieldname: string;
  label: string | null;
  fieldtype: ConnectorFieldType;
  /** Link: the target DocType. Select: newline-separated choices. */
  options: string | null;
  reqd: 0 | 1;
  description: string | null;
}

/** A stored secret never comes back — only whether there is one. */
export interface ConnectorPasswordValue {
  _type: "password";
  _set: boolean;
}

export type ConnectorValue = string | number | boolean | null | ConnectorPasswordValue;

export interface ConnectorConfig {
  fields: ConnectorField[];
  values: Record<string, ConnectorValue>;
}

export interface ConnectorTestResult {
  success: boolean;
  message: string;
}

export interface ConnectorRegistryRow {
  connector_id: string;
  connector_name: string;
  connector_app: string;
  connector_type: string;
  description: string | null;
  icon: string | null;
  settings_doctype: string | null;
  is_enabled: 0 | 1;
  connection_status: string | null;
  last_tested_at: string | null;
}

export function isPasswordValue(value: ConnectorValue): value is ConnectorPasswordValue {
  return typeof value === "object" && value !== null && (value as ConnectorPasswordValue)._type === "password";
}

/**
 * These endpoints `frappe.throw` rather than answering `{success: false}` — an
 * unregistered connector or a missing settings DocType is a deployment mistake,
 * not something a user can act on — so the message arrives in `_server_messages`
 * and has to be dug out to be worth showing.
 */
async function callMethod<T>(method: string, body?: Record<string, unknown>): Promise<T> {
  const res = await fetch(`/api/method/${method}`, {
    method: body ? "POST" : "GET",
    headers: body ? { "content-type": "application/json" } : undefined,
    body: body ? JSON.stringify(body) : undefined,
    cache: "no-store",
  });

  const text = await res.text();
  let payload: { message?: T; _server_messages?: string; exception?: string } | null = null;
  try {
    payload = JSON.parse(text) as typeof payload;
  } catch {
    payload = null;
  }

  if (!res.ok) throw new Error(serverMessage(payload) ?? `${method} failed (${res.status}).`);
  if (!payload || payload.message === undefined) throw new Error(`${method} returned nothing.`);
  return payload.message;
}

function serverMessage(payload: { _server_messages?: string; exception?: string } | null): string | null {
  if (!payload) return null;
  try {
    const messages = JSON.parse(payload._server_messages ?? "[]") as string[];
    const first = messages.map((entry) => JSON.parse(entry) as { message?: string }).find((entry) => entry.message);
    if (first?.message) return first.message.replace(/<[^>]+>/g, "");
  } catch {
    // fall through to the raw exception line
  }
  return payload.exception ?? null;
}

/** Every registered connector — for a settings index, or one connector's own status. */
export function fetchConnectors(): Promise<ConnectorRegistryRow[]> {
  return callMethod<ConnectorRegistryRow[]>("alaiy_os.api.connectors.get_all_connectors");
}

/** Field metadata plus current values for one connector's settings DocType. */
export function fetchConnectorConfig(connectorId: string): Promise<ConnectorConfig> {
  return callMethod<ConnectorConfig>(
    `alaiy_os.api.connectors.get_connector_config?connector_id=${encodeURIComponent(connectorId)}`,
  );
}

/**
 * Save, then run the connector's own connection test.
 *
 * The two are one call on purpose: settings that were never tested are the
 * reason a connector sits in "untested" forever. A Password field is only
 * written when a non-empty value is sent, so omitting it keeps the stored
 * secret — which is what lets a screen edit everything else without ever
 * holding the password.
 */
export function saveAndTestConnector(
  connectorId: string,
  values: Record<string, unknown>,
): Promise<ConnectorTestResult> {
  return callMethod<ConnectorTestResult>("alaiy_os.api.connectors.save_and_test", {
    connector_id: connectorId,
    values,
  });
}

/** Re-test what is already saved, and update the registry's connection status. */
export function testConnector(connectorId: string): Promise<ConnectorTestResult> {
  return callMethod<ConnectorTestResult>("alaiy_os.api.connectors.test_connector", {
    connector_id: connectorId,
  });
}
